Add 'auto_add_group_owner_membership' setting

bboro opened the pull request:
Case study:

Sometimes website may be creating entities programmatically, or just plain do not desire automatic membership creation for the group creator. In such cases, it's quite difficult to prevent this from happening due to the creation being in `og_entity_insert`.

Adding new global setting: 'Automatically add creators to the group' 

![image](https://user-images.githubusercontent.com/24897663/93938584-634d9b80-fd21-11ea-96af-9ce51d6345d5.png)

* When the setting is ON (default) membership is created
  automatically for group owners/creators.
* When the setting is OFF membership is not created automatically
  for group owners/creators.

Changelist:

* Added config schema changes
* Added default installation config changes
* Added update hook to update existing configurations
* Added the setting in the AdminSettingsForm
* Updated og_entity_insert to be aware of the new config setting
* Updated tests

bboro committed:
Add 'auto_add_owners' setting

* When the setting is ON (default) membership is created
  automatically for group owners/creators.
* When the setting is OFF membership is not created automatically
  for group owners/creators.

Changelist:

* Added config schema changes
* Added default installation config changes
* Added update hook to update existing configurations
* Added the setting in the AdminSettingsForm
* Updated og_entity_insert to be aware of the new config setting
* Updated tests

bboro committed:
Merge branch '8.x-1.x' into feature/auto-add-owners-setting

bboro commented:
This requirement was originally stemming from https://github.com/Gizra/jep-portal/issues/663 where we don't want to create memberships automatically for group owners.

bboro committed:
Fix the failing tests due to new settings

amitaibu commented on og_ui/src/Form/AdminSettingsForm.php:
```diff
@@ -133,6 +132,13 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       }
     }
 
+    $form['auto_add_owners'] = [
```
How about a more explicit name, like

```suggestion
    $form['auto_add_membership_to_group_owner'] = [
```

bboro replied:
Renamed to `auto_add_group_owner_membership`

bboro committed:
Rename auto_add_owners to auto_add_group_owner_membership

bboro commented:
Looks like some PHPCS failures, but looks like fixes are coming in another PR. I had a quick look at my test failures and seems like it's nothing introduced from my side, but I may have easily missed it.

pfrenssen commented:
I did a code review and everything looks good to me. Thanks for adding test coverage!

I did not test it out in the UI though, if someone can confirm this works as expected when setting the option in the UI this is good to go.

amitaibu committed:
Merge branch '8.x-1.x' into feature/auto-add-owners-setting

amitaibu commented:
I've confirmed manually the feature works

## Enabled

![image](https://user-images.githubusercontent.com/125707/95427334-0fad9580-0950-11eb-9d2f-98cb50e126d3.png)


## Disabled

![image](https://user-images.githubusercontent.com/125707/95427359-1a682a80-0950-11eb-8195-6d4d65c26e9c.png)


I think as a follow up we should show "No members in the group yet" message, as we've never had up until now empty groups. @bboro care to tackle this?

pfrenssen reviewed (approved):
Thanks for the test and the screenshots!

Approving, but we will have to wait for #700 to be in before merging this. We're affected by a test failure which is being fixed in that PR.

bboro commented:
@pfrenssen thanks for the review!

@amitaibu sure thing I'll make some time to work on this tomorrow. Question about the implementation: Since the members listing is a view, how should we handle the view configuration on existing installations? Do we just leave them be since they may have a custom empty message set already?

amitaibu commented:
> how should we handle the view configuration on existing installations?

You can change the existing one. As it was not (easily) possibly to have no-members, I don't think it's a problem.

bboro committed:
Add empty results text on members overview view optional install config

Also remove config hash from the view's installation config.

bboro committed:
Add update hook to add empty result text to Members overview view

bboro commented on og.install:
```diff
@@ -93,3 +94,48 @@ function og_update_8002() {
+    'tokenize' => FALSE,
+    'content' => [
+      'value' => 'No members in the group yet.',
+      'format' => 'basic_html',
```
Is it ok to assume basic_html is an existing format here? (Also same in the optional installation view config) 

If not, does anyone have any pointers on how this is usually handled? Perhaps I should use a different plugin/field?

amitaibu replied:
I don't think we need an upgrade path for Views config.

bboro replied:
> I don't think we need an upgrade path for Views config.

@amitaibu so we don't want to update existing installations and won't need this update hook? I'm not sure I understand exactly what you mean.

amitaibu replied:
Correct, no upgrade hook for Views

bboro replied:
Thanks! I'll remove

bboro replied:
FYI I'm also setting the empty text to unfiltered text in the installationg config yml instead of trying to force text formats

bboro committed:
Use filter_default_format() in update hook instead of assuming the format

bboro committed:
Remove the views update hook 8004

bboro committed:
Set the views empty text to unfiltered text instead of formatted

MPParsley committed:
Merge branch '8.x-1.x' into feature/auto-add-owners-setting

MPParsley commented:
#700 is fixed, updating branch.

MPParsley commented on og.install:
```diff
@@ -9,6 +9,7 @@ declare(strict_types = 1);
 
 use Drupal\Core\Field\BaseFieldDefinition;
 use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\views\Views;
```
Unused use statement:

```suggestion
```

MPParsley commented on tests/src/Kernel/GroupManagerSubscriptionTest.php:
```diff
@@ -117,6 +117,26 @@ public function testGroupManagerSubscription($group_has_owner, $membership_is_ov
+    ]);
+    $group->save();
+    $membership = $this->membershipManager->getMembership($group, $this->owner->id());
+    // Membership should be created only when auto_add_group_owner_membership is enabled.
```
Line exceeds 80 characters; contains 89 characters 

```suggestion
    // Membership should be created only when auto_add_group_owner_membership is
    // enabled.
```

MPParsley committed:
Apply suggestions from code review

MPParsley commented:
Fixed CS issues:

```
FILE: /home/travis/build/Gizra/og/og.install
----------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------
 12 | WARNING | [x] Unused use statement
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
FILE: ...me/travis/build/Gizra/og/tests/src/Kernel/GroupManagerSubscriptionTest.php
--------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
--------------------------------------------------------------------------------
 136 | WARNING | Line exceeds 80 characters; contains 89 characters
--------------------------------------------------------------------------------
```

MPParsley commented on tests/src/Kernel/GroupManagerSubscriptionTest.php:
```diff
@@ -176,4 +197,24 @@ public static function groupManagerSubscriptionProvider() {
+  public static function autoAddOwnersSettingProvider() {
+    return [
+      [
+        // Auto Add owners.
```
```suggestion
        // Auto add owners.
```

MPParsley commented on config/optional/views.view.og_members_overview.yml:
```diff
@@ -430,7 +428,18 @@ display:
+          admin_label: 'No results text'
+          empty: true
+          tokenize: false
+          content: 'No members in the group yet.'
```
```suggestion
          content: 'There are no members in this group.'
```

```suggestion
          content: 'There are no members that belong to this group.'
```

MPParsley reviewed (commented):
Test are green, just added a small grammar suggestion.

Should we make the default role configurable?

MPParsley committed:
Fix typo

bboro committed:
Merge remote-tracking branch 'origin/8.x-1.x' into feature/auto-add-owners-setting

bboro commented:
Rerolled against latest 8.x-1.x

bboro committed:
Restore the default config hash in Members overview view config

This is removed in another patch Gizra/og:#231 so restoring this
to avoid conflict.

MPParsley committed:
Merge branch '8.x-1.x' into feature/auto-add-owners-setting

MPParsley committed:
Fix grammar
